src/plugins/application_config/server/plugin.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Observable } from 'rxjs';
+import { first } from 'rxjs/operators';
+
+import {
+  PluginInitializerContext,
+  CoreSetup,
+  CoreStart,
+  Plugin,
+  Logger,
+  IScopedClusterClient,
+  SharedGlobalConfig,
+} from '../../../core/server';
+
+import {
+  ApplicationConfigPluginSetup,
+  ApplicationConfigPluginStart,
+  ConfigurationClient,
+} from './types';
+import { defineRoutes } from './routes';
+import { OpenSearchConfigurationClient } from './opensearch_config_client';
+
+export class ApplicationConfigPlugin
+  implements Plugin<ApplicationConfigPluginSetup, ApplicationConfigPluginStart> {
+  private readonly logger: Logger;
+  private readonly config$: Observable<SharedGlobalConfig>;
+
+  private configurationClient: ConfigurationClient;
+  private configurationIndexName: string;
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.logger = initializerContext.logger.get();
+    this.config$ = initializerContext.config.legacy.globalConfig$;
+    this.configurationIndexName = '';
+  }
+
+  private registerConfigurationClient(configurationClient: ConfigurationClient) {
+    this.logger.info('Register a configuration client.');
+
+    if (this.configurationClient) {
+      const errorMessage = 'Configuration client is already registered! Cannot register again!';
+      this.logger.error(errorMessage);
+      throw new Error(errorMessage);
+    }
+
+    this.configurationClient = configurationClient;
+  }
+
+  private getConfigurationClient(configurationClient: IScopedClusterClient): ConfigurationClient {
```

Comment:
Can we change the argument name to clusterClient instead of configurationClient? I was confused since we are trying to get configuration client, but had to pass a configuration client but its type is cluster client

Reply:
Good catch. Updated for better readability